File: back-end/src/main/java/nl/hva/backend/domain/models/game/board/location/locationImpl/RailRoad.java
package nl.hva.backend.domain.models.game.board.location.locationImpl;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import nl.hva.backend.domain.models.game.board.GameBoard;
import nl.hva.backend.domain.models.game.board.location.OwnableLocation;

import javax.persistence.Entity;

@Entity
@NoArgsConstructor
@Getter
@Setter
public class RailRoad extends OwnableLocation {
    private String name;
    private double rent;
    private double rentIfTwoAreOwned;
    private double rentIfThreeOwned;
    private double rentIfFourAreOwned;

    public RailRoad(
            long id,
            int position,
            String description,
            GameBoard gameBoard,
            double initialPurchasePrice,
            String name,
            double rent,
            double rentIfTwoAreOwned,
            double rentIfThreeOwned,
            double rentIfFourAreOwned
    ) {
        super(id, position, description, gameBoard, initialPurchasePrice);
        this.name = name;
        this.rent = rent;
        this.rentIfTwoAreOwned = rentIfTwoAreOwned;
        this.rentIfThreeOwned = rentIfThreeOwned;
        this.rentIfFourAreOwned = rentIfFourAreOwned;
    }
}

File: front-end/src/app/domain/models/game/board/location/locationImpl/RailRoad.ts
import {OwnableLocation} from "../OwnableLocation";

export class RailRoad extends OwnableLocation {
  private _name: string;
  private _rent: number;
  private _rentIfTwoAreOwned: number;
  private _rentIfThreeOwned: number;
  private _rentIfFourAreOwned: number;

  constructor(
    id: number,
    position: number,
    description: string,
    initialPurchasePrice: number,
    name: string,
    mortgageValue: number,
    rent: number,
    rentIfTwoAreOwned: number,
    rentIfThreeOwned: number,
    rentIfFourAreOwned: number
  ) {
    super(RailRoad.name, id, position, description, initialPurchasePrice, mortgageValue);
    this._name = name;
    this._rent = rent;
    this._rentIfTwoAreOwned = rentIfTwoAreOwned;
    this._rentIfThreeOwned = rentIfThreeOwned;
    this._rentIfFourAreOwned = rentIfFourAreOwned;
  }

  public static trueCopy(location: RailRoad): RailRoad {
    return new RailRoad(
      location.id,
      location.position,
      location.description,
      location.initialPurchasePrice,
      location.name,
      location.mortgageValue,
      location.rent,
      location.rentIfTwoAreOwned,
      location.rentIfThreeOwned,
      location.rentIfFourAreOwned,
    );
  }

  public override toJson(): object {
    return Object.assign(
      super.toJson(),
      {
        "initialPurchasePrice": this.initialPurchasePrice,
        "name": this.name,
        "mortgageValue": this.mortgageValue,
        "rent": this.rent,
        "rentIfTwoAreOwned": this.rentIfTwoAreOwned,
        "rentIfThreeOwned": this.rentIfThreeOwned,
        "rentIfFourAreOwned": this.rentIfFourAreOwned,
      }
    );
  }

  get name(): string {
    return this._name;
  }

  get rent(): number {
    return this._rent;
  }

  get rentIfTwoAreOwned(): number {
    return this._rentIfTwoAreOwned;
  }

  get rentIfThreeOwned(): number {
    return this._rentIfThreeOwned;
  }

  get rentIfFourAreOwned(): number {
    return this._rentIfFourAreOwned;
  }

  override toString(): string {
    return this.name;
  }
}
